File: packages/db/src/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes } from 'node:crypto';

/**
 * Envelope encryption for mailbox secrets.
 *
 * Per-user data key wrapping the refresh token; DEKs wrapped by a KEK held
 * outside the database. A stolen dump yields nothing readable, which is the
 * only claim worth making about a box that reads your mail.
 *
 * ── Why not libsodium ──────────────────────────────────────────────────────
 *
 * §02 names `sodium-native` for `crypto_secretbox`, and its reason is the right
 * one: "audited primitives, no room to choose a mode wrongly." That reason is
 * satisfied here by AES-256-GCM from Node's own crypto — an audited AEAD, one
 * mode, no options to get wrong, and no dependency at all. The alternatives
 * both cost something this deployment cannot afford: `sodium-native` is a
 * native addon that needs a compiler in the image and a rebuild on every Node
 * bump on an ARM box, and `libsodium-wrappers` ships an ESM build whose
 * internal import does not resolve under Node's resolver.
 *
 * The one real difference is nonce size: XSalsa20's 24 random bytes tolerate
 * far more messages per key than GCM's 12 before birthday-bound collisions
 * matter. It does not bite here — a DEK seals one secret per mailbox, and a KEK
 * wraps one DEK per user, so the count is in the single digits against a bound
 * of 2^32. If that ever stops being true, the fix is a fresh DEK per message,
 * not a bigger nonce. decisions.md B18.
 */

const ALGORITHM = 'aes-256-gcm';
const NONCE_BYTES = 12;
const KEY_BYTES = 32;
const TAG_BYTES = 16;

export interface Sealed {
  /** Ciphertext with the 16-byte auth tag appended. */
  ciphertext: Buffer;
  nonce: Buffer;
}

/** Kept for API symmetry with the async libsodium version the spec assumed. */
export async function initCrypto(): Promise<void> {
  // Node's crypto needs no initialisation; this exists so callers written
  // against the spec's `await initCrypto()` keep working.
}

export function loadKek(env = process.env.LOOP_KEK): Buffer {
  if (!env) {
    throw new Error(
      'LOOP_KEK is not set. Generate one with: node -e "console.log(require(\'crypto\').randomBytes(32).toString(\'base64\'))"',
    );
  }
  const key = Buffer.from(env, 'base64');
  if (key.length !== KEY_BYTES) throw new Error(`LOOP_KEK must decode to 32 bytes, got ${key.length}`);
  return key;
}

export function generateDek(): Buffer {
  return randomBytes(KEY_BYTES);
}

export function seal(plaintext: Buffer | string, key: Buffer): Sealed {
  if (key.length !== KEY_BYTES) throw new Error('key must be 32 bytes');
  const nonce = randomBytes(NONCE_BYTES);
  const cipher = createCipheriv(ALGORITHM, key, nonce);
  const body = Buffer.concat([
    cipher.update(typeof plaintext === 'string' ? Buffer.from(plaintext, 'utf8') : plaintext),
    cipher.final(),
  ]);
  return { ciphertext: Buffer.concat([body, cipher.getAuthTag()]), nonce };
}

export function open(sealed: Sealed, key: Buffer): Buffer {
  if (key.length !== KEY_BYTES) throw new Error('key must be 32 bytes');
  if (sealed.ciphertext.length < TAG_BYTES) throw new Error('ciphertext is too short to carry a tag');

  const body = sealed.ciphertext.subarray(0, sealed.ciphertext.length - TAG_BYTES);
  const tag = sealed.ciphertext.subarray(sealed.ciphertext.length - TAG_BYTES);
  const decipher = createDecipheriv(ALGORITHM, key, sealed.nonce);
  decipher.setAuthTag(tag);
  // A wrong key throws here rather than returning plausible bytes, which is the
  // whole reason this is an AEAD and not a raw cipher.
  return Buffer.concat([decipher.update(body), decipher.final()]);
}

export function openUtf8(sealed: Sealed, key: Buffer): string {
  return open(sealed, key).toString('utf8');
}

/** DEK sealed with the KEK. What goes in `mailbox_accounts.dek_wrapped`. */
export function wrapDek(dek: Buffer, kek: Buffer = loadKek()): Sealed {
  return seal(dek, kek);
}

export function unwrapDek(wrapped: Sealed, kek: Buffer = loadKek()): Buffer {
  return open(wrapped, kek);
}

/**
 * KEK rotation: re-wrap every DEK in one transaction. `scripts/rotate-kek.ts`
 * drives it and a test covers it — because the runbook promises it works, and
 * an untested rotation is a promise rather than a procedure.
 */
export function rewrapDek(wrapped: Sealed, oldKek: Buffer, newKek: Buffer): Sealed {
  return seal(unwrapDek(wrapped, oldKek), newKek);
}

const SECRET_KEY_PATTERN = /token|password|secret|authorization|cookie|credential|refresh|access_key/i;

/**
 * The log serialiser. "Plaintext secrets exist only inside the connector
 * process, only for the length of one call, and are never placed in a variable
 * that a logger can reach" — this is the second half of that sentence.
 */
export function redact<T>(value: T): T {
  const walk = (node: unknown): unknown => {
    if (Array.isArray(node)) return node.map(walk);
    if (Buffer.isBuffer(node)) return '[buffer]';
    if (node && typeof node === 'object') {
      const out: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(node as Record<string, unknown>)) {
        out[k] = SECRET_KEY_PATTERN.test(k) ? '[redacted]' : walk(v);
      }
      return out;
    }
    return node;
  };
  return walk(value) as T;
}

File: backend/loop/google/crypto.py
"""Envelope encryption for the one secret this system holds.

A refresh token to a mailbox is the whole of the trust the user extends, so it
is sealed under a per-mailbox data key, and that key is sealed under a key held
outside the database. A stolen dump yields nothing readable — which is the only
claim worth making about a box that reads your mail.

AES-256-GCM, and the reasoning carries over from the reference unchanged: an
audited AEAD, one mode, nothing to choose wrongly, and no native dependency to
rebuild on every interpreter bump. The nonce is twelve bytes, which is small
enough to matter at volume and does not: a data key seals one secret per
mailbox and the key wraps one data key per user, so the count is in single
digits against a bound of 2^32.

The format is byte-compatible with the TypeScript one — the tag appended to the
ciphertext, the nonce beside it — because both implementations read the same
rows for as long as both run.
"""

import base64
import os
from dataclasses import dataclass
from typing import Final

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

NONCE_BYTES: Final = 12
KEY_BYTES: Final = 32


@dataclass(frozen=True, slots=True)
class Sealed:
    """Ciphertext with the sixteen-byte tag appended, and its nonce."""

    ciphertext: bytes
    nonce: bytes


def load_kek(value: str | None = None) -> bytes:
    key_material = value if value is not None else os.environ.get("LOOP_KEK")
    if not key_material:
        raise ValueError(
            "LOOP_KEK is not set. Generate one with: "
            'python -c "import os,base64;print(base64.b64encode(os.urandom(32)).decode())"'
        )
    key = base64.b64decode(key_material)
    if len(key) != KEY_BYTES:
        raise ValueError(f"LOOP_KEK must decode to 32 bytes, got {len(key)}")
    return key


def generate_dek() -> bytes:
    return os.urandom(KEY_BYTES)


def seal(plaintext: bytes, key: bytes) -> Sealed:
    nonce = os.urandom(NONCE_BYTES)
    return Sealed(AESGCM(key).encrypt(nonce, plaintext, None), nonce)


def open_sealed(sealed: Sealed, key: bytes) -> bytes:
    """Raises rather than returning plausible bytes on the wrong key.

    That is the AEAD tag doing its job, and it is why a mis-set `LOOP_KEK` is a
    loud failure at the first sync rather than a mailbox that mysteriously
    stops authenticating.
    """
    plaintext: bytes = AESGCM(key).decrypt(sealed.nonce, sealed.ciphertext, None)
    return plaintext


def wrap_dek(dek: bytes, kek: bytes | None = None) -> Sealed:
    return seal(dek, kek or load_kek())


def unwrap_dek(wrapped: Sealed, kek: bytes | None = None) -> bytes:
    return open_sealed(wrapped, kek or load_kek())


def rewrap_dek(wrapped: Sealed, old_kek: bytes, new_kek: bytes) -> Sealed:
    """KEK rotation, one data key at a time.

    `scripts/rotate_kek.py` drives it over every row in one transaction, and a
    test covers it — because the runbook promises rotation works, and an
    untested rotation is a promise rather than a procedure.

    Note what this does *not* touch: the sealed refresh token itself. Rotating
    the key-encryption key re-wraps the data keys and leaves every ciphertext
    where it is, which is the whole reason for the envelope.
    """
    return seal(unwrap_dek(wrapped, old_kek), new_kek)
